lib/em-http/http_encoding.rb

```diff
@@ -51,7 +51,10 @@ def encode_request(method, uri, query, proxy)
 
       # Non CONNECT proxies require that you provide the full request
       # uri in request header, as opposed to a relative path.
-      query = uri.join(query) if proxy
+      # Don't modify the header with CONNECT proxies. It's unneeded and will
+      # cause 400 Bad Request errors with many standard setups.
+      is_connect = proxy && (proxy[:type] == :http || proxy[:type].nil?) && uri.scheme == "https"
+      query = uri.join(query) if proxy && !is_connect
```

Comment:
I'm a bit dubious of this one.. as it seems we'd fix one thing, break another? If I remember correctly, Apache/Squid require that you send the full request URL in connect, which is why this logic is in place.. but nginx errors out? 

To make things even more interesting, the spec does seem to indicate that we should only send the hostname and port number: http://tools.ietf.org/html/rfc7231#section-4.3.6

That aside.. In terms of implementation, I think we can make this cleaner: pass connopts object to this method and check for "http_proxy?" instead of duplicating logic.